File: qbopt/analysis/memoryssa.py
"""Conservative memory SSA over MIR, rebuilt after a body changes.

Stores, calls and barriers define a single memory state; loads use that state.
The clobber walker skips stores proven disjoint by MIR alias analysis.
Raised call write effects participate in the same alias queries as stores.
Calls without write metadata and barriers remain conservative.
"""

from dataclasses import dataclass
from enum import StrEnum

from qbopt.analysis import effects, loops, pointerfacts
from qbopt.model import mir


def _unknown_write(op: mir.Op) -> bool:
    return effects.unmodeled_write(op) or op.floating is not None or op.kind is mir.Kind.FCHECK


class Kind(StrEnum):
    LIVE = "live-on-entry"
    USE = "use"
    DEF = "def"
    PHI = "phi"


@dataclass(frozen=True, slots=True)
class Site:
    block: int
    index: int


@dataclass(frozen=True, slots=True)
class Access:
    id: int
    kind: Kind
    block: int | None = None
    site: Site | None = None
    defining: int | None = None
    # None denotes the invocation edge into the entry block.
    incoming: tuple[tuple[int | None, int], ...] = ()


@dataclass(frozen=True, slots=True)
class MemorySSA:
    live: Access
    accesses: tuple[Access, ...]
    sites: dict[Site, Access]
    phis: dict[int, Access]
    operations: dict[Site, mir.Op]
    pointers: pointerfacts.Offsets

    def at(self, site: Site) -> Access:
        return self.sites[site]

    def clobbers(
        self, site: Site, memory: mir.MemRef, dgroup: frozenset[int] = frozenset(),
    ) -> frozenset[int]:
        """Possible nearest writes before a site, including live-on-entry.

        Walk every phi input. A visited set closes cycles without treating
        the backedge as evidence that memory is unchanged. An empty result
        means no reachable source was found, not a reusable memory value.
        This identifies memory states, not a dominating scalar definition;
        forwarding consumers must establish value availability separately.
        """
        return self._frontier(site, memory, dgroup)

    def unchanged(
        self, earlier: Site, later: Site, memory: mir.MemRef,
        dgroup: frozenset[int] = frozenset(),
    ) -> bool:
        """Whether a dominating earlier read's memory state still applies.

        The caller must establish dominance and equal addresses. Stop at
        the earlier memory version, rejecting any possibly aliasing write
        on the way, including writes carried by loop backedges.
        """
        boundary = self.at(earlier).defining
        return boundary is not None and self._frontier(later, memory, dgroup, boundary) == frozenset({boundary})

    def _frontier(
        self, site: Site, memory: mir.MemRef, dgroup: frozenset[int], boundary: int | None = None,
        edge: int | None = None,
        edge_memory: mir.MemRef | None = None,
    ) -> frozenset[int]:
        accesses = {access.id: access for access in self.accesses}
        pending = [self.at(site).defining]
        seen: set[int] = set()
        found: set[int] = set()
        while pending:
            current = pending.pop()
            if current is None or current in seen:
                continue
            seen.add(current)
            if current == boundary:
                found.add(current)
                continue
            access = accesses[current]
            match access.kind:
                case Kind.LIVE:
                    found.add(current)
                case Kind.PHI:
                    pending.extend(value for parent, value in access.incoming
                                   if edge is None or access.block != site.block or parent == edge)
                case Kind.DEF:
                    op = self.operations[access.site]
                    queried = edge_memory if edge_memory is not None and access.block != site.block else memory
                    if _unknown_write(op) or any(
                        mir.overlapping(queried, store, dgroup) and not self.pointers.disjoint(queried, store)
                        for store in op.stores
                    ):
                        found.add(current)
                    else:
                        pending.append(access.defining)
                case Kind.USE:
                    pending.append(access.defining)
        return frozenset(found)

    def available_on_edge(
        self, earlier: Site, later: Site, predecessor: int, memory: mir.MemRef,
        dgroup: frozenset[int] = frozenset(),
        *, edge_memory: mir.MemRef | None = None,
    ) -> bool:
        """An earlier load/store still supplies these bytes on one incoming edge.

        The caller proves equal addresses and that the scalar provider dominates
        the predecessor. Only the destination's memory phi is edge-selected;
        other intervening joins still require agreement along every path.
        A translated address applies outside the destination; its prefix must
        still be checked against the original phi-based address.
        """
        source = self.at(earlier)
        boundary = source.id if source.kind is Kind.DEF else source.defining
        return boundary is not None and self._frontier(later, memory, dgroup, boundary, predecessor, edge_memory) == frozenset({boundary})


def built(body: mir.MirBody) -> MemorySSA:
    """Wire block entries first, then eliminate identity memory phis.

Preallocating entries handles backedges without iterative guesses about
memory versions. Entry blocks with backedges retain an invocation input.
"""
    live = Access(0, Kind.LIVE)
    entries = {block.at: index + 1 for index, block in enumerate(body.blocks)}
    sites: dict[Site, Access] = {}
    outgoing: dict[int, int] = {}
    next_id = len(entries) + 1
    for block in body.blocks:
        current = entries[block.at]
        for index, op in enumerate(block.ops):
            defines = bool(op.stores or _unknown_write(op))
            if not (op.loads or defines):
                continue
            kind = Kind.DEF if defines else Kind.USE
            site = Site(block.at, index)
            sites[site] = Access(next_id, kind, block.at, site, current)
            if kind is Kind.DEF:
                current = next_id
            next_id += 1
        outgoing[block.at] = current

    predecessors = loops.predecessors(body.blocks)
    incoming = {
        block.at: tuple((pred, outgoing[pred]) for pred in sorted(predecessors[block.at]))
        + (((None, live.id),) if block.at == body.entry or not predecessors[block.at] else ())
        for block in body.blocks
    }
    replacements: dict[int, int] = {}

    def resolved(value: int) -> int:
        while value in replacements:
            value = replacements[value]
        return value

    changed = True
    while changed:
        changed = False
        for block, entry in entries.items():
            if entry in replacements:
                continue
            values = {resolved(value) for _, value in incoming[block]} - {entry}
            if len(values) <= 1:
                replacements[entry] = next(iter(values), live.id)
                changed = True

    phis = {
        block: Access(entry, Kind.PHI, block, incoming=tuple(
            (pred, resolved(value)) for pred, value in incoming[block]
        ))
        for block, entry in entries.items()
        if entry not in replacements
    }
    sites = {
        site: Access(access.id, access.kind, access.block, site, resolved(access.defining))
        for site, access in sites.items()
    }
    accesses = tuple(sorted((live, *phis.values(), *sites.values()), key=lambda access: access.id))
    operations = {
        Site(block.at, index): op
        for block in body.blocks for index, op in enumerate(block.ops)
        if Site(block.at, index) in sites
    }
    return MemorySSA(live, accesses, sites, phis, operations, pointerfacts.offsets(body))

File: src/analysis/memoryssa.rs
//! Port of `qbopt/analysis/memoryssa.py`.

// ---- early port (agent B) ----

use std::collections::BTreeSet;

use indexmap::IndexMap;

use super::{effects, loops, pointerfacts, regions};
use crate::model::mir::{self, MemRef, MirBody, Op};

fn _unknown_write(op: &Op) -> bool {
    effects::unmodeled_write(op) || op.floating.is_some() || op.kind == mir::Kind::Fcheck
}

#[derive(Clone, Copy, Debug, Eq, Hash, PartialEq)]
pub(crate) enum Kind {
    Live,
    Use,
    Def,
    Phi,
}

#[derive(Clone, Copy, Debug, Eq, Hash, Ord, PartialEq, PartialOrd)]
pub(crate) struct Site {
    pub block: i64,
    pub index: usize,
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub(crate) struct Access {
    pub id: i64,
    pub kind: Kind,
    pub block: Option<i64>,
    pub site: Option<Site>,
    pub defining: Option<i64>,
    /// `None` denotes the invocation edge into the entry block.
    pub incoming: Vec<(Option<i64>, i64)>,
}

#[derive(Clone, Debug)]
#[allow(dead_code)] // `live` and `phis` are read by the rest of the module
pub(crate) struct MemorySSA<'a> {
    pub live: Access,
    pub accesses: Vec<Access>,
    pub sites: IndexMap<Site, Access>,
    pub phis: IndexMap<i64, Access>,
    pub operations: IndexMap<Site, &'a Op>,
    pub pointers: pointerfacts::Offsets<'a>,
}

impl MemorySSA<'_> {
    pub(crate) fn at(&self, site: Site) -> &Access {
        &self.sites[&site]
    }

    /// Possible nearest writes before a site, including live-on-entry.
    pub(crate) fn clobbers(&self, site: Site, memory: &MemRef, dgroup: &BTreeSet<i64>) -> BTreeSet<i64> {
        self._frontier(site, memory, dgroup, None, None, None)
    }

    fn _frontier(
        &self,
        site: Site,
        memory: &MemRef,
        _dgroup: &BTreeSet<i64>,
        boundary: Option<i64>,
        edge: Option<i64>,
        edge_memory: Option<&MemRef>,
    ) -> BTreeSet<i64> {
        let accesses = self.accesses.iter().map(|access| (access.id, access)).collect::<IndexMap<_, _>>();
        let mut pending = vec![self.at(site).defining];
        let mut seen = BTreeSet::new();
        let mut found = BTreeSet::new();
        while let Some(current) = pending.pop() {
            let Some(current) = current.filter(|current| !seen.contains(current)) else {
                continue;
            };
            seen.insert(current);
            if Some(current) == boundary {
                found.insert(current);
                continue;
            }
            let access = accesses[&current];
            match access.kind {
                Kind::Live => {
                    found.insert(current);
                }
                Kind::Phi => pending.extend(
                    access
                        .incoming
                        .iter()
                        .filter(|(parent, _)| edge.is_none() || access.block != Some(site.block) || *parent == edge)
                        .map(|(_, value)| Some(*value)),
                ),
                Kind::Def => {
                    let op = self.operations[&access.site.expect("a def has a site")];
                    let queried = match edge_memory {
                        Some(edge_memory) if access.block != Some(site.block) => edge_memory,
                        _ => memory,
                    };
                    // `mir.overlapping` passes `dgroup` as a layout, which is
                    // not a `module.Group`: it is Rust's `layout=None`.  An
                    // endpoint Rust cannot represent may overlap.
                    if _unknown_write(op)
                        || op.stores.iter().any(|store| {
                            regions::overlapping(queried, store, None, None, None).unwrap_or(true)
                                && !self.pointers.disjoint(queried, store)
                        })
                    {
                        found.insert(current);
                    } else {
                        pending.push(access.defining);
                    }
                }
                Kind::Use => pending.push(access.defining),
            }
        }
        found
    }
}

/// Wire block entries first, then eliminate identity memory phis.
pub(crate) fn built(body: &MirBody) -> MemorySSA<'_> {
    let live = Access {
        id: 0,
        kind: Kind::Live,
        block: None,
        site: None,
        defining: None,
        incoming: Vec::new(),
    };
    let entries = body
        .blocks
        .iter()
        .enumerate()
        .map(|(index, block)| (block.at, index as i64 + 1))
        .collect::<IndexMap<_, _>>();
    let mut sites = IndexMap::<Site, Access>::new();
    let mut outgoing = IndexMap::<i64, i64>::new();
    let mut next_id = entries.len() as i64 + 1;
    for block in &body.blocks {
        let mut current = entries[&block.at];
        for (index, op) in block.ops.iter().enumerate() {
            let defines = !op.stores.is_empty() || _unknown_write(op);
            if !(!op.loads.is_empty() || defines) {
                continue;
            }
            let kind = if defines { Kind::Def } else { Kind::Use };
            let site = Site { block: block.at, index };
            sites.insert(
                site,
                Access {
                    id: next_id,
                    kind,
                    block: Some(block.at),
                    site: Some(site),
                    defining: Some(current),
                    incoming: Vec::new(),
                },
            );
            if kind == Kind::Def {
                current = next_id;
            }
            next_id += 1;
        }
        outgoing.insert(block.at, current);
    }

    let predecessors = loops::predecessors(&body.blocks);
    let empty = BTreeSet::new();
    let incoming = body
        .blocks
        .iter()
        .map(|block| {
            let preds = predecessors.get(&block.at).unwrap_or(&empty);
            let mut edges = preds
                .iter()
                .map(|pred| (Some(*pred), outgoing[pred]))
                .collect::<Vec<_>>();
            if block.at == body.entry || preds.is_empty() {
                edges.push((None, live.id));
            }
            (block.at, edges)
        })
        .collect::<IndexMap<_, _>>();
    let mut replacements = IndexMap::<i64, i64>::new();

    let resolved = |replacements: &IndexMap<i64, i64>, mut value: i64| {
        while let Some(next) = replacements.get(&value) {
            value = *next;
        }
        value
    };

    let mut changed = true;
    while changed {
        changed = false;
        for (block, entry) in &entries {
            if replacements.contains_key(entry) {
                continue;
            }
            let mut values = incoming[block]
                .iter()
                .map(|(_, value)| resolved(&replacements, *value))
                .collect::<BTreeSet<_>>();
            values.remove(entry);
            if values.len() <= 1 {
                replacements.insert(*entry, values.first().copied().unwrap_or(live.id));
                changed = true;
            }
        }
    }

    let phis = entries
        .iter()
        .filter(|(_, entry)| !replacements.contains_key(*entry))
        .map(|(block, entry)| {
            (
                *block,
                Access {
                    id: *entry,
                    kind: Kind::Phi,
                    block: Some(*block),
                    site: None,
                    defining: None,
                    incoming: incoming[block]
                        .iter()
                        .map(|(pred, value)| (*pred, resolved(&replacements, *value)))
                        .collect(),
                },
            )
        })
        .collect::<IndexMap<_, _>>();
    let sites = sites
        .into_iter()
        .map(|(site, access)| {
            let defining = access.defining.map(|value| resolved(&replacements, value));
            (
                site,
                Access {
                    site: Some(site),
                    defining,
                    ..access
                },
            )
        })
        .collect::<IndexMap<_, _>>();
    let mut accesses = std::iter::once(live.clone())
        .chain(phis.values().cloned())
        .chain(sites.values().cloned())
        .collect::<Vec<_>>();
    accesses.sort_by_key(|access| access.id);
    let operations = body
        .blocks
        .iter()
        .flat_map(|block| {
            block
                .ops
                .iter()
                .enumerate()
                .map(move |(index, op)| (Site { block: block.at, index }, op))
        })
        .filter(|(site, _)| sites.contains_key(site))
        .collect::<IndexMap<_, _>>();
    MemorySSA {
        live,
        accesses,
        sites,
        phis,
        operations,
        pointers: pointerfacts::offsets(body),
    }
}
